src/udev/udev-builtin-blkid.c

```diff
@@ -234,11 +241,80 @@ static int probe_superblocks(blkid_probe pr) {
         return blkid_do_safeprobe(pr);
 }
 
+static int read_loopback_backing_inode(
+                sd_device *dev,
+                int fd,
+                dev_t *ret_devno,
+                ino_t *ret_inode,
+                char **ret_fname) {
+
+        _cleanup_free_ char *fn = NULL;
+        struct loop_info64 info;
+        const char *name;
+        int r;
+
+        assert(dev);
+        assert(fd >= 0);
+        assert(ret_devno);
+        assert(ret_inode);
+        assert(ret_fname);
+
+        /* Retrieves various fields of the current loopback device backing file, so that we can ultimately
+         * use it to create stable symlinks to loopback block devices, based on what they are backed by. We
+         * pick up inode/device as well as file name field. Note that we pick up the "lo_file_name" field
+         * here, which is an arbitrary free-form string provided by userspace. We do not return the sysfs
+         * attribute loop/backing_file here, because that is directly accessible from udev rules anyway. And
+         * sometimes, depending on context, it's a good thing to return the string userspace can freely pick
+         * over the string automatically generated by the kernel. */
+
+        r = sd_device_get_sysname(dev, &name);
+        if (r < 0)
+                return r;
+
+        if (!startswith(name, "loop"))
+                goto notloop;
+
+        if (ioctl(fd, LOOP_GET_STATUS64, &info) < 0) {
+                if (ERRNO_IS_NOT_SUPPORTED(errno))
+                        goto notloop;
+
+                return -errno;
+        }
+
+#if HAVE_VALGRIND_MEMCHECK_H
+        VALGRIND_MAKE_MEM_DEFINED(&info, sizeof(info));
+#endif
+
+        if (isempty((char*) info.lo_file_name) ||
+            strnlen((char*) info.lo_file_name, sizeof(info.lo_file_name)-1) == sizeof(info.lo_file_name)-1)
```

Comment:
kernel insists on trailing NUL, hence if we read this from the kernel, then there'll always be a trailing NUl byte. We assume truncation if the string uses all space before that hence. Now strictly speaking this doesn't mean truncation, it could just be a non-truncated string that uses the full length possible, but I think we should better be safe than sorry here, and only accept strings as non-truncated that are definitely not truncated. 

Or in other words: if the maximum length is actually used here, we don't know if it's truncated or not, hence we assume the worst.

Will add comment.

Reply:
Note that "losetup" and similar tools on display suffix all strings set this way that use the max length with "*" to indicate there might be more coming. Here we take inspiration and just are tiny bit tougher.

Ultimately, whatever we do, we only have 63 chars here, which is not enough to allow arbitrary paths or just filenames, hence the qeustion is mostly how early we refuse to use the data.

Reply:
> Or in other words: if the maximum length is actually used here, we don't know if it's truncated or not,

I may be confused, but I think the buffer is fully used without railing NUL when
```c
strnlen((char*) info.lo_file_name, sizeof(info.lo_file_name)) == sizeof(info.lo_file_name)
```
and `-1` is not necessary, no?

Reply:
if you write a string into "lo_file_name" that uses all space available so that a trailing NUL byte cannot be added, and then configure that onto a loop device, then the kernel takes the string and writes a NUL byte onto the last char. hence the longest sensible string that makes sense to pass to the kernel is `sizeof(info.lo_file_name)-1`, because if you pass one that is `sizeof(info.lo_file_name)` long, then the kernel will simply chop off the last byte.

See here: https://github.com/torvalds/linux/blob/master/drivers/block/loop.c#L987

hence when reading back the string via
LOOP_GET_STATUS64 it really only makes sense to look at the `sizeof(info.lo_file_name)-1` chars of it, since the final byte is going to be zero anyway.